Change Playbook <=> Gear Association to Many to Many

LeftoverCurry opened the pull request:
### **Issue:**
Gear was originally set up to belong to an individual playbook, but in reality the same gear could belong to many different playbooks.

### **Resolution:**
* Created a `has_many: :through` ActiveRecord association for each table and set up a through table with the model PlaybookGear.  
* Removed `playbook_id` from the Gear table.
* Restructured the currently built out Playbook seeds to no longer reference the gear.
* Restructured and built out the Gear seeds to contain a static list of available gear and their respective playbooks.
* Created a new tag type called 'keywords' that can distinguish between sets of gear within a playbook such as The Professional, who can choose one 'serious' weapon and two 'normal' weapons.  Currently also includes means of transport if the playbook has them, keyworded as 'transport'. The new seed file will populate these.
* Added `after` block to Gear seeds to ensure it runs after the playbook is created.
* Updated spec files to reflect the new architecture.
* Added Categories to PlaybookGear through table so that we can differentiate when a Playbook has different categories of gear.

### **Notes:**
* You were asking on stream briefly about the [spec file for hunters_improvements](spec/controllers/hunters_improvements_controller_spec.rb) and why I changed it to include `to_s`.   Previously, the test for the hunters_improvements JSON error message would pass because it was set with a matcher that included the exact text it would respond with ('does not match improvement playbook: ').  After adding a name to the factory it would no longer pass, as it was attempting to find 'does not match improvement playbook: ' within an array that contained only one index that read 'does not match improvement playbook: The Unspecified'. I converted that object into a string so that it checked for the same thing (that the array contained the error text) but was not dependant on hard coding in the name of the factory, should we ever decide to change it, have multiple factories, or use Faker.


resolves #96

LeftoverCurry commented:
### **Issue:**
Gear was originally set up to belong to an individual playbook, but in reality the same gear could belong to many different playbooks.

### **Resolution:**
* Created a `has_many: :through` ActiveRecord association for each table and set up a through table with the model PlaybookGear.  
* Restructured and built out the Gear seeds to contain a static list of available gear.
* Restructured the currently built out Playbook seeds to reference the gear ID and build the through table.
* Added `after` block to Playbook seeds to ensure it runs after the gear is created.
* Removed `playbook_id` from the Gear table.
* Used the currently existing and unused `description` column in the Gear table to provide data to make selections when a Playbook has multiple sets of gear, such as The Professional, who can choose one 'serious' weapon and two 'normal' weapons.
* Updated spec files to reflect the new architecture.

### **Notes:**
* This will require a `db:reset`, as the Gear references only work if they are in the correct order in the database since they are selected by ID.
resolves #96

ChaelCodes commented on db/migrate/20200526200555_add_gear_to_playbook.rb:
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
```
Can you remove this migration from the commit history?

LeftoverCurry replied:
Removed in commit 6ef0a4a

ChaelCodes commented on db/migrate/20200526205042_remove_playbook_id_from_gears.rb:
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
```
And then we can remove this one!

ChaelCodes commented on db/migrate/20200526230159_remove_available_gears_from_playbooks.rb:
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class RemoveAvailableGearsFromPlaybooks < ActiveRecord::Migration[6.0]
```
And then we can remove this one!

LeftoverCurry replied:
Removed in commit 6ef0a4a

ChaelCodes commented on db/seeds/playbooks/crooked.seeds.rb:
```diff
@@ -274,57 +274,21 @@
+  # Baseball bat (1-harm hand) - 8
+  # Submachinegun (2-harm close reload area) - 73
+  # Assault rifle (3-harm close/far area) - 5
+  [1, 2, 3, 62, 46, 12, 8, 73, 5].each do |gear_id|
```
This list of ids will be different in each environment, especially production, which already gear loaded. Name and Harm would be a better way to associate this.

LeftoverCurry replied:
We discussed this elsewhere, but just to have a record of what was decided in case we need to look back over it:
Unfortunately, searching by harm and name won't work, because we have entries that share name and harm.  The solution we came up with was to populate the through table through a reference in gear to what playbooks they belong in.  Initial plan is create a new column referencing the playbook names in the gear table, then rewrite the seed files to populate in that manner.

LeftoverCurry replied:
Addressed in commits d886634 and f886cf9

ChaelCodes commented on spec/models/playbook_spec.rb:
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
```
Thank you for creating this spec! There was one in minitest that needs to be deleted, not that it has been migrated to RSpec.
Please remove this file:
https://github.com/ChaelCodes/HuntersKeepers/blob/10e66ea18cf29db1e2e06b8a65ee2eb9ca63a967/test/models/playbook_test.rb#L15
Thank you for migrating the playbook spec to RSpec!

LeftoverCurry replied:
Removed with commit 692f05e

ChaelCodes commented on spec/models/playbook_gear_spec.rb:
```diff
@@ -0,0 +1,18 @@
+
+  it 'correctly pulls gear from a playbook' do
+    playbook.gears << gear
+    expect(playbook.gears.ids).to include(gear.id)
```
Rails has a helper you can use here:
```suggestion
    expect(playbook.gear_ids).to include(gear.id)
```
Totally optional, but it's a nice piece of syntactic sugar.

LeftoverCurry replied:
Poured some sugar on it.  Fixed with commit 3c872f3.

ChaelCodes commented on spec/factories/playbooks.rb:
```diff
@@ -1,16 +1,8 @@
 # frozen_string_literal: true
 
-# == Schema Information
```
Can we restore this schema info?

LeftoverCurry replied:
There appeared to be several files that needed documentation updates.  I ran the annotate gem and included them all in commit dc1a1f1.  I think it might be worth considering adding a task to the rakefile to automate that process whenever a new migration is run. 

From the gem documentation:
![Screen Shot 2020-06-01 at 11 02 51 AM](https://user-images.githubusercontent.com/20918543/83422429-857c8000-a3f7-11ea-8753-e9aa3fe4ee9b.png)

LeftoverCurry replied:
Added issue #132 to automate documentation.

ChaelCodes commented on db/seeds/playbooks/professional.seeds.rb:
```diff
@@ -214,78 +214,35 @@
+  # Big knife (1-harm hand) - 12
+
+  # serious or normal can be searched by the description column for each gear
+  [31, 21, 4, 36, 69, 37, 74, 2, 3, 46, 62, 12].each do |gear_id|
```
This list of ids will be different in each environment, especially production, which already has gear loaded. Name and Harm would be a better way to associate this.

LeftoverCurry replied:
We discussed this elsewhere, but just to have a record of what was decided in case we need to look back over it:
Unfortunately, searching by harm and name won't work, because we have entries that share name and harm. The solution we came up with was to populate the through table through a reference in gear to what playbooks they belong in. Initial plan is create a new column referencing the playbook names in the gear table, then rewrite the seed files to populate in that manner.

LeftoverCurry replied:
Addressed in commits d886634 and f886cf9

ChaelCodes commented on db/seeds/playbook.seeds.rb:
```diff
@@ -1,136 +1,138 @@
-end
+  true weirdness — but I’ll investigate them all to find it.'
+   }].each do |playbook|
+    Playbook.find_or_create_by(name: playbook[:name]).update!(playbook)
```
Why was this changed to `find_or_create_by`? I don't think the playbook is valid on creation, because the description isn't included in the attributes.

LeftoverCurry replied:
Reverting and will revisit once new gear seeding is completed.

ChaelCodes commented on app/models/playbook.rb:
```diff
@@ -4,17 +4,22 @@
 class Playbook < ApplicationRecord
   has_many :improvements
-  has_many :gears
+  has_many :playbook_gears
```
I think our dependent option should be `destroy`.

LeftoverCurry replied:
Indeed.  Resolved with commit: 977bbfe

ChaelCodes commented on app/controllers/hunters_controller.rb:
```diff
@@ -18,8 +18,10 @@ def index
+  # Available_gear contains all gear available to the Hunter's current playbook
   def show
     @available_improvements = params[:available_improvements]
+    @available_gear = @hunter.playbook.gears
```
Where do we use `@available_gear`?

LeftoverCurry replied:
We do not.  This was added to address the last bullet point on the issue, which ended up not being needed.  Removed in commit c6538bd.

LeftoverCurry commented:
### Correcting Issue With File Seeding Using Index Number

**Ref Commits d886634, f886cf9, cba031e, 4b79985**

* Added a playbooks key to the hash and populated it with which playbooks the gear would belong to using the instance variables declared in the playbooks seed.
* Added a few description keywords I missed on the first run through.
* Broke off the playbooks values from the generation of the gears and used it to populate `Gears.playbooks`.  This prevents the need to create a new column that would duplicate information.
* Removed the `after:` from playbooks and added it to gear, as the seed order had reversed.
* Removed references to gear in the individual playbook's seed files, as we are no longer populating from there.

LeftoverCurry commented:
### Regarding cba031e
I could not test the app unless I basically adopted the changes from PR #128, so if there are any merge conflicts, let me know and I'll get it cleaned up.

ChaelCodes commented on app/controllers/hunters_controller.rb:
```diff
@@ -18,6 +18,7 @@ def index
   # @param :available_improvements [Boolean] Toggles whether or not to include a list of improvements available to
   #   the hunter
   # @see Hunter#available_improvements
+  # Available_gear contains all gear available to the Hunter's current playbook
```
Comment leftover.

LeftoverCurry replied:
Fixed with commit d7d6308.

ChaelCodes replied:
I'm still seeing this in the final file chantges.

LeftoverCurry replied:
should be fixed as of 59ce133.

ChaelCodes commented on app/controllers/hunters_controller.rb:
```diff
@@ -32,7 +33,8 @@ def new
   #
   # @param id [Integer] id of the Hunter to be edited
-  def edit() end
+  # Available_gear contains all gear available to the Hunter's current playbook
```
Comment leftover.

LeftoverCurry replied:
Fixed with commit d7d6308.

ChaelCodes replied:
Same here, I'm still seeing this in the final file changes.

LeftoverCurry replied:
Should be fixed as of 59ce133.

ChaelCodes commented on app/models/gear.rb:
```diff
@@ -9,14 +9,13 @@
 #  created_at  :datetime         not null
 #  updated_at  :datetime         not null
 #
-# Gear is equipment that the Hunter can use.
```
This is our top-level comment. I think we still want it. We can get rid of the playbook reference though.

LeftoverCurry replied:
Resolved with commit 817d7cc.  Tested with the annotate gem and top level documentation was not removed.

ChaelCodes commented on app/models/hunter.rb:
```diff
@@ -19,7 +19,6 @@
 #  playbook_id :bigint
 #  user_id     :bigint
 #
-# The character appearing in the mystery
```
That's a top-level comment that we want to keep.

LeftoverCurry replied:
Resolved with commit 817d7cc. Tested with the annotate gem and top level documentation was not removed.

ChaelCodes commented on app/models/playbook.rb:
```diff
@@ -10,11 +10,13 @@
   has_many :ratings, dependent: :destroy
+
+  validates :name, presence: true
+  validates :description, presence: true
```
Like the extra validations! ✨

ChaelCodes commented on db/seeds/gear.seeds.rb:
```diff
@@ -1,103 +1,101 @@
+  ].each do |gear|
+    tag_list = gear.delete(:tags)
+    playbooks = gear.delete(:playbooks)
+    g = Gear.create!(gear)
```
What if the gear already exists? We run our seed files every deploy. I believe this would constantly recreate gear in our database.

LeftoverCurry replied:
Corrected to `find_or_initialize_by` since the save is called later in the code.  Resolved in commit 3501c33.

ChaelCodes commented on spec/controllers/hunters_improvements_controller_spec.rb:
```diff
@@ -30,7 +30,7 @@
   let(:valid_session) { {} }
 
-  before(:each) do
+  before do
```
Was this a linter fix? If so, in another PR, I'd like to configure the Rubocop rules to recommend the more specific option.

LeftoverCurry replied:
It was a local issue that I have fixed.

LeftoverCurry replied:
Resolved with commit d1c055e.

ChaelCodes commented on db/seeds/gear.seeds.rb:
```diff
@@ -1,103 +1,101 @@
+    { name: '9mm', harm: 2, tags: 'close, loud', description: 'normal, modern, practical', playbooks: [@crooked, @flake, @initiate, @monstrous, @professional, @spooky, @wronged] },
+    { name: 'Assault Rifle', harm: 3, tags: 'far, area, loud, reload', description: 'serious', playbooks: [@professional] },
+    { name: 'Assault Rifle', harm: 3, tags: 'close/far, area',  playbooks: [@crooked] },
+    { name: 'Assault Rifle', harm: 3, tags: 'close, area, loud, reload', description: 'signature', playbooks: [@wronged] },
```
What are these descriptions for?

LeftoverCurry replied:
Resolved with commit b8fcd8c.

ChaelCodes commented on db/seeds/gear.seeds.rb:
```diff
@@ -1,103 +1,101 @@
+    { name: 'Axe', harm: 2, tags: 'hand, messy', playbooks: [@initiate], description: 'old-fashioned' },
+    { name: 'Baseball Bat', harm: 1, tags: 'hand', playbooks: [@crooked] },
+    { name: 'Baseball Bat', harm: 2, tags: 'hand, innocuous, messy', playbooks: [@mundane] },
+    { name: 'Bicycle', tags: '', playbooks: [@mundane], description: 'mundane_transport' },
```
This one has underscores???

LeftoverCurry replied:
Resolved with commit b8fcd8c

LeftoverCurry commented on db/seeds/gear.seeds.rb:
```diff
@@ -1,103 +1,103 @@
+    { name: 'Truck', tags: '', keywords: 'transport', playbooks: [@professional] },
+    { name: 'Van', tags: '', keywords: 'transport', playbooks: [@mundane, @professional] },
+    { name: "Watchman's Flashlight", harm: 1, tags: 'hand', playbooks: [@flake] },
+    { name: 'Weighted Gloves/Brass Knuckles', harm: 1, tags: 'hand', playbooks: [@flake] }
```
So, the issue here is that the playbooks don't know what categories they want, so it's always going to pick whatever category comes first (or last, if you're overwriting).  So if you're populating, say, a 9mm, that has multiple categories it can fit into (Normal, Modern, and Practical, in this case), all PlaybookGear entries for 9mm will categorized as normal, even if the referenced playbook does not have that category.  My fix for this would be to include an acts_as_taggable on Playbook tagging the categories that the playbook itself includes and then using that in the seed file to distinguish which categories it's looking for.

LeftoverCurry replied:
Added below.

LeftoverCurry commented on app/models/playbook.rb:
```diff
@@ -10,11 +12,14 @@
+
+  validates :name, presence: true
+  validates :description, presence: true
+  acts_as_taggable_on :categories
```
Added a category tag to playbooks here so that the playbooks knew which categories to seed with.

LeftoverCurry commented on app/models/playbook_gear.rb:
```diff
@@ -0,0 +1,18 @@
+# Through table model for Playbook <=> Gear
+class PlaybookGear < ApplicationRecord
+  belongs_to :playbook
+  belongs_to :gear
```
Added migration.

LeftoverCurry commented on db/seeds/playbook.seeds.rb:
```diff
@@ -13,7 +14,8 @@
 and magic… well… it wasn’t so different from the underworld I
-already knew. It was easy to find an angle, just like before.'
+already knew. It was easy to find an angle, just like before.',
+   categories: 'transport'
```
Added categories here.  I do feel like they would fit better in the individual playbooks, but I did it here as a work around solution until the individual playbooks are built out.

LeftoverCurry commented on db/seeds/playbook.seeds.rb:
```diff
@@ -113,7 +120,12 @@
+    categories = playbook.delete(:categories)
+    pb.category_list = categories
+  end
+  pb.update(playbook)
```
See comment above.  We needed a way for the playbooks to know which categories they had so that if a gear could be in multiple categories it wouldn't just pick whatever category was first.  I built out tags to indicate them and saved them in the category_list for the playbook.

LeftoverCurry commented on spec/controllers/gears_controller_spec.rb:
```diff
@@ -7,7 +7,6 @@
   # adjust the attributes here as well.
   let(:valid_attributes) {
     {
```
We no longer need this attribute and it causes the tests to fail.

LeftoverCurry commented on spec/factories/playbook_gears.rb:
```diff
@@ -2,18 +2,16 @@
-  # end
+FactoryBot.define do
+  factory :playbook_gear do
+  end
```
Removed Minitest and migrated to RSpec.

LeftoverCurry commented on spec/models/playbook_gear_spec.rb:
```diff
@@ -0,0 +1,29 @@
+  it 'correctly pulls playbooks from gear' do
+    playbook.gears << gear
+    expect(gear.playbook_ids).to include(playbook.id)
+  end
```
Added a model test for PlaybookGear.

LeftoverCurry commented on .gitignore:
```diff
@@ -37,3 +37,7 @@ yarn-debug.log*
+
+# Ignore VSCode debugger config
+/.vscode
+/bin/rspec
```
Allows a local user to use VSCode's debugger.

LeftoverCurry commented on app/controllers/gears_controller.rb:
```diff
@@ -72,6 +72,6 @@ def set_gear
   # Never trust parameters from the scary internet, only allow the white list through.
   def gear_params
-    params.require(:gear).permit(:name, :description, :harm, :armor, :playbook_id, :tag_list)
+    params.require(:gear).permit(:name, :description, :harm, :armor, :tag_list)
```
The Playbook relationship is now handled under PlaybookGear.

LeftoverCurry commented on app/models/gear.rb:
```diff
@@ -1,5 +1,13 @@
+# have different numbers of selectable gear, such as with The Flake, who can
+# have one 'normal' weapon and two 'hidden' weapons.  Keywords can also be used
+# to reference modes of transport using the keyword 'transport'.
+
```
All user generated top level comments have been moved above the documentation generated by the annotate gem, as it deletes anything between it and the class.

LeftoverCurry commented on app/views/gears/index.html.erb:
```diff
@@ -23,7 +23,6 @@
         <td><%= gear.armor %></td>
         <td><%= gear.tag_list %></td>
-        <td><%= link_to_playbook(gear) %></td>
         <td><%= link_to 'Show', gear %></td>
```
Gear is no longer associated with an individual playbook.

LeftoverCurry commented on db/migrate/20200526160952_create_playbook_gears.rb:
```diff
@@ -0,0 +1,10 @@
+    create_table :playbook_gears do |t|
+      t.references :playbook, null: false, foreign_key: true
+      t.references :gear, null: false, foreign_key: true
+
```
I added a migration below to add it.

LeftoverCurry commented on db/seeds/playbook.seeds.rb:
```diff
@@ -134,3 +146,4 @@
 @pararomantic = Playbook.find_by name: 'The Pararomantic'
 @searcher = Playbook.find_by name: 'The Searcher'
 @initiate = Playbook.find_by name: 'The Initiate'
+@playbooks = Playbook.all
```
Added a general helper here for the Playbook Collection.

LeftoverCurry commented on spec/factories/playbooks.rb:
```diff
@@ -12,5 +12,7 @@
 FactoryBot.define do
   factory :playbook do
+    name { 'The Unspecified' }
+    description { 'A mysterious wonderer' }
```
Needed a name to check for in one of the tests.

ChaelCodes commented:
You built a genuinely good PR here. Looking at the differences between gear on the different playbooks though, I think it'd be better to maintain a 1-to-many instead of many-to-many. Sorry!
